src/Client.ts

```diff
@@ -95,24 +103,55 @@ export default class Client extends discord.Client {
     // Enable dispatcher.
     this.on('interactionCreate', (interaction) => {
       if (interaction instanceof CommandInteraction) {
-        dispatch(interaction, commands);
+        dispatch(interaction, commands, this);
       }
 
       // FIXME figure out a button/select menu api that
-      /*
       if (interaction instanceof ButtonInteraction) {
-        const handler = client.buttonListeners.get(interaction.customId);
+        const handler = this.buttonListeners.get(interaction.customId);
 
         if (!handler) {
           return;
         }
 
         // Run handler.
-        handler(interaction);
+        handler(interaction, this);
       }
-      */
     });
   }
+
+  /**
+   * Generates a discord.js `MessageActionRow[]` that can be used in a
+   * message reply as the `components` argument. Allows use of `onClick` and
+   * `onSelect` by autogenerating and registering IDs.
+   *
+   * @param ui Either a single `UIComponent` or a 1D or 2D array of `UIComponent`s
+   * @returns a generated `MessageActionRow[]`
+   */
+  registerUI(
+    ui: UIComponent | UIComponent[] | UIComponent[][]
+  ): MessageActionRow[] {
```

Comment:
Since this returns an action row of buttons, how would you add other non-homogenous components like menus, or link buttons (which aren't permitted to have custom IDs)?

Reply:
I haven't tested link buttons but they should already work because if you don't provide an `onClick` property, then IDs aren't generated.

Reply:
This PR initially was going to have select menu support, but I haven't fully read the docs on select menus so I just pushed up this part. Basically, the plan is to provide a `DispatchSelect` class similar to the `DispatchButton` class with an `onSelect` method. I can easily just add another `instanceof` check and case to the setup function to figure out what is what. The `UIComponent` type is intended to be a union type of all the current and future UI components.

Reply:
Ok I tested this with a link button and it did not work. Investigating now...

Reply:
Another thing to keep in mind here is that you can only have one select menu per row, while you can have 5 buttons per row

Reply:
True, I will add validation for these things to avoid errors.